### PR TITLE
Use store with k8s slim types

### PR DESCRIPTION
**Pull request by aanm**
Add the ability to use our own types in cache.Store. This will help decrease the memory consumption of Cilium for large clusters or small clusters with a large amount of pods.

```release-note
Reduce memory consumption in k8s watchers
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7433)

### CODE REVIEW

File: pkg/k8s/factory_functions.go [hunk 1]
```diff
@@ -215,3 +216,126 @@ func EqualV1Namespace(ns1, ns2 *v1.Namespace) bool {
...
+	*ingress = v1beta1.Ingress{}
+	return ing
+}
+
+func ConvertToCNPWithStatus(obj interface{}) interface{} {
+	cnp, ok := obj.(*cilium_v2.CiliumNetworkPolicy)
+	if !ok {
+		return nil
+	}
+	slimCNP := &types.SlimCNP{
```
Comment (ianvernon) [inline pkg/k8s/factory_functions.go:290]: Can you explain what is being eliminated from the original CNP here?

Comment (aanm) [reply pkg/k8s/factory_functions.go:290]: Nothing is.

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]
PR is difficult to review but looks good

**Comment by aanm** [comment]
test-me-please

File: pkg/k8s/factory_functions.go [hunk 2]
```diff
@@ -229,3 +230,139 @@ func EqualV1Namespace(ns1, ns2 *v1.Namespace) bool {
...
+		},
+	}
+	*cnp = cilium_v2.CiliumNetworkPolicy{}
+	return slimCNP
+}
+
+// ConvertToPod converts a *v1.Pod into a *types.Pod.
+// WARNING calling this function will set *all* fields of the given Pod as
+// empty.
+func ConvertToPod(obj interface{}) interface{} {
```
Comment (ianvernon) [inline pkg/k8s/factory_functions.go:320]: Can you explain why the fields need to be set to empty? Is this so the old state can be GC'd?

**Review by ianvernon — APPROVED** [review]

**Comment by aanm** [comment]
test-missed-k8s

**Comment by aanm** [comment]
test-me-please

**Pull request merged**